File: site/src/lib/catalog.mjs
export const TESTED_VERSION = '1.2.93.667.g7b5cc0ce';
export const LIVE_CATALOG = 'https://raw.githubusercontent.com/LoaderSpot/table/main/table/versions.json';
export const LEGACY_CATALOG = 'https://raw.githubusercontent.com/LoaderSpot/LoaderSpot/main/versions.json';
export const LINUX_PACKAGES = 'https://repository.spotify.com/dists/stable/non-free/binary-amd64/Packages';
export const REPOSITORY = 'RobyRew/BlockTheSpot-Installer';
export const PERMANENT = { x64: '/SpotifyFullSetupX64.exe', arm64: '/SpotifyFullSetupARM64.exe', x86: '/SpotifyFullSetup.exe' };
const versionPattern = /^1\.\d{1,5}\.\d{1,5}\.\d{1,8}\.g[0-9a-f]{8,40}$/i;
const targets = [
  ['win', 'x64', 'windows', 'x64', 'win32-x86_64', 'exe'],
  ['win', 'x86', 'windows', 'x86', 'win32-x86', 'exe'],
  ['win', 'arm64', 'windows', 'arm64', 'win32-arm64', 'exe'],
  ['mac', 'intel', 'macos', 'x64', 'osx-x86_64', 'tbz'],
  ['mac', 'arm64', 'macos', 'arm64', 'osx-arm64', 'tbz'],
  ['linux', 'amd64', 'linux', 'x64', '', 'deb'],
];

export function compareVersions(a, b) {
  const left = a.split('.').slice(0, 4).map(Number);
  const right = b.split('.').slice(0, 4).map(Number);
  for (let i = 0; i < 4; i++) if (left[i] !== right[i]) return left[i] - right[i];
  return 0;
}

function isoDate(value) {
  if (typeof value !== 'string' || !/^\d{2}\.\d{2}\.\d{4}$/.test(value)) return null;
  const result = value.split('.').reverse().join('-');
  const date = new Date(result);
  return Number.isFinite(+date) && date.toISOString().startsWith(result) ? result : null;
}

export function sourceFor(raw, fullVersion, platform, architecture) {
  if (!versionPattern.test(fullVersion)) return null;
  let url;
  try { url = new URL(raw); } catch { return null; }
  if (url.protocol !== 'https:' || url.username || url.password || url.port || url.search || url.hash) return null;
  const target = targets.find(([, , p, a]) => p === platform && a === architecture);
  if (!target) return null;
  const [, , , , directory, format] = target;
  const stem = platform === 'windows' ? `spotify_installer-${fullVersion}` : `spotify-autoupdate-${fullVersion}`;
  if (url.hostname === 'upgrade.scdn.co' && platform !== 'linux') {
    const prefix = `/upgrade/client/${directory}/${stem}-`;
    if (url.pathname.startsWith(prefix) && new RegExp(`^\\d+\\.${format}$`).test(url.pathname.slice(prefix.length)))
      return { url: url.href, kind: 'official', label: 'Spotify CDN' };
  }
  if (url.hostname === 'repository.spotify.com' && platform === 'linux' &&
      url.pathname === `/pool/non-free/s/spotify-client/spotify-client_${fullVersion}_amd64.deb`)
    return { url: url.href, kind: 'official', label: 'Spotify repository' };
  // Spotify's permanent full installers always serve the current build; applyOfficial attaches
  // one only to the build whose ETag the watcher last saw there.
  if (url.hostname === 'download.scdn.co' && platform === 'windows' && url.pathname === PERMANENT[architecture])
    return { url: url.href, kind: 'official', label: 'Spotify (current build)' };
  if (url.hostname === 'github.com' && platform === 'windows' &&
      new RegExp(`^/${REPOSITORY}/releases/download/[A-Za-z0-9._-]+/spotify_installer-${fullVersion.replace(/\./g, '\\.')}-${architecture}\\.exe$`).test(url.pathname))
    return { url: url.href, kind: 'archive', label: 'GitHub archive' };
  if (url.hostname === 'loadspot.amd64fox1.workers.dev') {
    const suffix = platform === 'macos' && architecture === 'x64' ? 'x86_64' : architecture;
    const file = platform === 'linux' ? `spotify-client_${fullVersion}_amd64.deb` : `${stem}-${suffix}.${format}`;
    if (url.pathname === `/download/${file}`) return { url: url.href, kind: 'mirror', label: 'LoadSpot mirror' };
  }
  return null;
}

export function normalizeCatalog(catalog) {
  if (!catalog || typeof catalog !== 'object' || Array.isArray(catalog)) throw new Error('Expected a version catalog object');
  const entries = [];
  for (const [version, entry] of Object.entries(catalog)) {
    if (!entry || typeof entry !== 'object' || !versionPattern.test(entry.fullversion ?? '') ||
        entry.fullversion.split('.').slice(0, 4).join('.') !== version ||
        (entry.buildType && entry.buildType.toLowerCase() !== 'release')) continue;
    for (const [group, key, platform, architecture, , format] of targets) {
      const asset = entry[group]?.[key] ?? entry.links?.[group]?.[key];
      const url = typeof asset === 'string' ? asset : asset?.url;
      const source = sourceFor(url, entry.fullversion, platform, architecture);
      if (!source) continue;
      entries.push({
        id: `${entry.fullversion}-${platform}-${architecture}`,
        version, fullVersion: entry.fullversion, platform, architecture, format,
        date: isoDate(asset?.date), size: Number.isSafeInteger(asset?.size) && asset.size > 0 ? asset.size : null,
        tested: platform === 'windows' && architecture === 'x64' && entry.fullversion.toLowerCase() === TESTED_VERSION,
        sources: [source],
      });
    }
  }
  return entries;
}

export function parseLinuxPackages(text) {
  const catalog = {};
  for (const paragraph of text.trim().split(/\r?\n\s*\r?\n/)) {
    const fields = Object.fromEntries(paragraph.split(/\r?\n/).filter(line => /^[A-Za-z0-9-]+: /.test(line))
      .map(line => [line.slice(0, line.indexOf(':')), line.slice(line.indexOf(':') + 2)]));
    const full = fields.Version?.replace(/^\d+:/, '');
    if (fields.Package !== 'spotify-client' || fields.Architecture !== 'amd64' || !versionPattern.test(full ?? '')) continue;
    const url = new URL(fields.Filename, 'https://repository.spotify.com/').href;
    if (!sourceFor(url, full, 'linux', 'x64')) continue;
    const version = full.split('.').slice(0, 4).join('.');
    catalog[version] = { fullversion: full, linux: { amd64: { url, size: Number(fields.Size) } } };
  }
  return normalizeCatalog(catalog);
}

export function mergeCatalogs(...catalogs) {
  const merged = new Map();
  for (const entry of catalogs.flat()) {
    const previous = merged.get(entry.id);
    const sources = new Map([...(previous?.sources ?? []), ...entry.sources].map(source => [source.url, source]));
    merged.set(entry.id, { ...entry, date: entry.date ?? previous?.date ?? null, size: entry.size ?? previous?.size ?? null,
      ...(entry.sha256 ?? previous?.sha256 ? { sha256: entry.sha256 ?? previous.sha256 } : {}),
      sources: [...sources.values()].sort((a, b) => a.kind.localeCompare(b.kind) || a.url.localeCompare(b.url)) });
  }
  const platforms = ['windows', 'macos', 'linux'];
  const architectures = ['x64', 'arm64', 'x86'];
  return [...merged.values()].sort((a, b) => compareVersions(b.version, a.version) ||
    platforms.indexOf(a.platform) - platforms.indexOf(b.platform) ||
    architectures.indexOf(a.architecture) - architectures.indexOf(b.architecture) || a.id.localeCompare(b.id));
}

// Permanent Spotify links and the apt repository never expire; the CI archive is a copy taken from
// them; LoadSpot mirrors are maintained; old Spotify CDN links have expired and come last.
const preference = ['Spotify (current build)', 'Spotify repository', 'GitHub archive', 'LoadSpot mirror', 'Spotify CDN'];
export function orderedSources(entry, kind = 'all') {
  return entry.sources.filter(source => kind === 'all' || source.kind === kind)
    .toSorted((a, b) => preference.indexOf(a.label) - preference.indexOf(b.label));
}
export function selectSource(entry, kind = 'all') { return orderedSources(entry, kind)[0]; }
export function sourceNote(source) {
  if (source.kind === 'archive') return 'Copied from Spotify by CI · hash listed';
  if (source.kind === 'mirror') return 'Community hosted';
  return source.label === 'Spotify CDN' ? 'Archived link · may have expired' : 'Direct download';
}

/**
 * Overlays the release watcher's observations (site/data/official.json) on the catalog: adds the
 * SHA-256 CI computed from Spotify's own file, the CI archive copy when one exists, and, for the
 * build currently behind a permanent URL, that URL with its ETag. Nothing here is persisted into
 * catalog.json, so a permanent link never outlives the build it pointed at.
 */
export function applyOfficial(entries, official) {
  if (!official?.builds?.length) return entries;
  const current = new Set(Object.values(official.watched ?? {}).map(watched => watched.etag));
  const observed = [];
  for (const build of official.builds) {
    if (!versionPattern.test(build.fullVersion ?? '') || !/^[0-9a-f]{64}$/.test(build.sha256 ?? '')) continue;
    const sources = [];
    if (current.has(build.etag) && sourceFor(build.url, build.fullVersion, build.platform, build.architecture)?.label === 'Spotify (current build)')
      sources.push({ url: build.url, kind: 'official', label: 'Spotify (current build)', etag: build.etag });
    const archive = build.archive && sourceFor(build.archive, build.fullVersion, build.platform, build.architecture);
    if (archive?.kind === 'archive') sources.push(archive);
    // The update service's http_prefix is Spotify's versioned link; it expires, but it is kept like
    // the historical CDN links so the exact official path of every build stays on record.
    const prefix = build.updateService?.httpPrefix && sourceFor(build.updateService.httpPrefix, build.fullVersion, build.platform, build.architecture);
    if (prefix?.label === 'Spotify CDN') sources.push(prefix);
    observed.push({
      id: `${build.fullVersion}-${build.platform}-${build.architecture}`,
      version: build.fullVersion.split('.').slice(0, 4).join('.'), fullVersion: build.fullVersion,
      platform: build.platform, architecture: build.architecture, format: 'exe',
      date: typeof build.lastModified === 'string' ? build.lastModified.slice(0, 10) : null,
      size: Number.isSafeInteger(build.size) && build.size > 0 ? build.size : null,
      tested: build.platform === 'windows' && build.architecture === 'x64' && build.fullVersion.toLowerCase() === TESTED_VERSION,
      sha256: build.sha256, sources,
    });
  }
  return mergeCatalogs(entries, observed);
}

export function filterCatalog(entries, { query = '', platform = 'all', architecture = 'all', source = 'all', sort = 'newest' } = {}) {
  const search = query.trim().toLowerCase();
  const result = entries.filter(entry => (!search || entry.fullVersion.toLowerCase().includes(search)) &&
    (platform === 'all' || entry.platform === platform) && (architecture === 'all' || entry.architecture === architecture) &&
    !!selectSource(entry, source));
  return sort === 'oldest' ? result.toReversed() : result;
}

// Installer feed: every Windows x64 build. `url` keeps the LoadSpot mirror (the stable link) so
// older parsers still work. `official` is Spotify's own link: the permanent full installer with
// its `etag` while that build is current, otherwise the historical upgrade.scdn.co link, which
// Spotify has usually expired. `archive` is the CI copy and `sha256` the hash CI took from Spotify's file.
export function windowsFeed(entries) {
  // A build known only by its hash (no longer current, not mirrored, not archived) has nothing to offer the installer.
  return Object.fromEntries(entries.filter(entry => entry.platform === 'windows' && entry.architecture === 'x64' && entry.sources.length > 0).map(entry => {
    const find = label => entry.sources.find(source => source.label === label);
    const current = find('Spotify (current build)');
    const official = current ?? find('Spotify CDN');
    const mirror = find('LoadSpot mirror')?.url;
    const archive = find('GitHub archive')?.url;
    return [entry.version, { fullversion: entry.fullVersion, win: { x64: {
      url: mirror ?? archive ?? official.url,
      ...(official && (mirror || archive) ? { official: official.url } : {}),
      ...(current ? { etag: current.etag } : {}),
      ...(archive ? { archive } : {}),
      ...(entry.sha256 ? { sha256: entry.sha256 } : {}),
      ...(entry.date ? { date: entry.date.split('-').reverse().join('.') } : {}), size: entry.size ?? 0,
    } } }];
  }));
}

export function formatSize(bytes) { return bytes ? `${(bytes / 1048576).toFixed(1)} MB` : '—'; }
export const platformNames = { windows: 'Windows', macos: 'macOS', linux: 'Linux' };

File: src/BlockTheSpot.Core/Catalog.cs
using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace BlockTheSpot.Core;

public static class Sources
{
    public const string Repository = "https://github.com/RobyRew/BlockTheSpot-Installer";
    public static readonly Uri Catalog = new("https://raw.githubusercontent.com/LoaderSpot/table/main/table/versions.json");
    public static readonly Uri CatalogApi = new("https://robyrew.github.io/BlockTheSpot-Installer/api/v1/windows-x64.json");
    public static readonly Uri Config = new("https://github.com/Nuzair46/BlockTheSpot/releases/latest/download/config.ini");
    public static readonly Uri Chrome = new("https://github.com/Nuzair46/BlockTheSpot/releases/latest/download/chrome_elf.dll");
    public static readonly Uri Block = new("https://github.com/Nuzair46/BlockTheSpot/releases/latest/download/blockthespot.dll");
    // Spotify's only permanent, unsigned installer URL. It always serves the current version.
    public static readonly Uri LatestSpotify = new("https://download.scdn.co/SpotifyFullSetupX64.exe");
    public static readonly Uri LatestRelease = new("https://api.github.com/repos/RobyRew/BlockTheSpot-Installer/releases/latest");
    // Versioned installers: Spotify hands the client a signed, short-lived upgrade.scdn.co link
    // for the current version only (spclient desktop-update/v2, authenticated). Older links expire
    // with HTTP 403, so a catalog entry keeps Spotify's link as the first attempt and the LoadSpot
    // mirror as the fallback. Every download is still checked against Spotify's Authenticode signature.
    public const string MirrorHost = "loadspot.amd64fox1.workers.dev";
    public const string UpgradeHost = "upgrade.scdn.co";
    // The release watcher (site/scripts/watch-official.mjs) downloads each new build from
    // download.scdn.co, records its SHA-256 and ETag, and can attach the file to this release.
    public const string ArchiveRepository = "RobyRew/BlockTheSpot-Installer";
}

/// <summary>
/// One installable Spotify build. <see cref="Url"/> is tried first, then <see cref="Archive"/>, then <see cref="Mirror"/>.
/// <see cref="Sha256"/> is the hash CI took from Spotify's own file; <see cref="ETag"/> binds the permanent URL to this build.
/// </summary>
public sealed record SpotifyChoice(string? FullVersion, Uri Url, string? Date = null, long Size = 0, bool Recommended = false, Uri? Mirror = null, bool Custom = false,
    Uri? Archive = null, string? Sha256 = null, string? ETag = null)
{
    public static SpotifyChoice Latest { get; } = new(null, Sources.LatestSpotify);
    public bool IsLatest => FullVersion is null;
    public string Title => FullVersion ?? "Latest official Spotify";
    public string Badge => Recommended ? "Tested" : IsLatest ? "Official" : Custom ? "Custom" : "Untested";
    public string Source => SourceOf(Url);
    public string Detail => string.Join(" · ", new[] { IsLatest ? "Current release" : Date, Size > 0 ? $"{Size / 1048576d:F0} MiB" : null, Source,
        Sha256 is null ? null : "SHA-256" }.Where(s => !string.IsNullOrWhiteSpace(s)));
    public IEnumerable<Uri> Urls => new[] { Url, Archive, Mirror }.Where(u => u is not null).Distinct()!;
    public static string SourceOf(Uri url) => SpotifyVersions.IsOfficial(url) ? "Spotify" : url.Host == "github.com" ? "GitHub archive" : "LoadSpot mirror";
    public override string ToString() => Title;
}

public sealed record CatalogResult(string? Tested, IReadOnlyList<SpotifyChoice> Choices, SpotifyChoice Selected, string? Warning = null);

public static partial class SpotifyVersions
{
    [GeneratedRegex(@"^1\.\d+\.\d+\.\d+(?:\.g[0-9a-fA-F]+)?$", RegexOptions.CultureInvariant)]
    private static partial Regex VersionPattern();
    [GeneratedRegex(@"^1\.\d{1,5}\.\d{1,5}\.\d{1,8}\.g[0-9a-fA-F]{8,40}$", RegexOptions.CultureInvariant)]
    private static partial Regex FullVersionPattern();

    public static Version Parse(string text)
    {
        if (!VersionPattern().IsMatch(text)) throw new InvalidDataException($"Unrecognized Spotify version: {text}");
        if (!Version.TryParse(string.Join('.', text.Split('.').Take(4)), out var version))
            throw new InvalidDataException($"Unrecognized Spotify version: {text}");
        return version;
    }

    public static string MinimumFromConfig(string text) =>
        text.Split('\n').Select(line => line.Trim().TrimStart('﻿'))
            .Where(line => line.StartsWith(';')).Select(line => line[1..].Trim())
            .FirstOrDefault(line => VersionPattern().IsMatch(line))
        ?? throw new InvalidDataException("BlockTheSpot's configuration does not contain a supported Spotify version.");

    public static void ValidateInstalled(string installed, string minimum, SpotifyChoice? pinned)
    {
        var actual = Parse(installed);
        if (actual < Parse(minimum))
            throw new InvalidOperationException($"Spotify {installed} is older than the supported minimum {minimum}. Install a supported version before patching.");
        if (pinned?.FullVersion is { } target && actual != Parse(target))
            throw new InvalidOperationException($"Spotify {installed} is installed, but {target} was selected. Setup did not finish as expected; no patch was applied.");
    }

    private static bool IsPlainHttps(Uri url) =>
        url.Scheme == "https" && url.IsDefaultPort && url.UserInfo.Length == 0 && url.Query.Length == 0 && url.Fragment.Length == 0;

    /// <summary>A link on Spotify's own hosts: the permanent full installer or a versioned x64 upgrade package.</summary>
    public static bool IsOfficial(Uri url) => (IsPlainHttps(url) && url == Sources.LatestSpotify) || VersionFromOfficial(url) is not null;

    public static bool IsMirror(Uri url, string version) =>
        IsPlainHttps(url) && url.Host == Sources.MirrorHost && url.AbsolutePath == $"/download/spotify_installer-{version}-x64.exe";

    /// <summary>A CI copy on this repository's releases, named after the exact build.</summary>
    public static bool IsArchive(Uri url, string version) =>
        IsPlainHttps(url) && url.Host == "github.com" && string.Equals(VersionFromArchive(url), version, StringComparison.OrdinalIgnoreCase);

    public static bool IsCatalogDownload(Uri url, string version) =>
        IsMirror(url, version) || IsArchive(url, version) || string.Equals(VersionFromOfficial(url), version, StringComparison.OrdinalIgnoreCase);

    public static Uri MirrorFor(string fullVersion) => new($"https://{Sources.MirrorHost}/download/spotify_installer-{fullVersion}-x64.exe");

    private static string? VersionFromArchive(Uri url)
    {
        if (!IsPlainHttps(url) || url.Host != "github.com") return null;
        var match = Regex.Match(url.AbsolutePath, $@"^/{Regex.Escape(Sources.ArchiveRepository)}/releases/download/[A-Za-z0-9._-]+/spotify_installer-(1\.\d+\.\d+\.\d+\.g[0-9a-fA-F]+)-x64\.exe$", RegexOptions.CultureInvariant);
        return match.Success ? match.Groups[1].Value : null;
    }

    // Spotify's versioned links carry a signed, 30-day ?fauth= token; it is the only query string accepted anywhere.
    private static bool IsSignedUpgrade(Uri url) =>
        url.Scheme == "https" && url.IsDefaultPort && url.UserInfo.Length == 0 && url.Fragment.Length == 0 && url.Host == Sources.UpgradeHost
        && Regex.IsMatch(url.Query, @"^\?fauth=[A-Za-z0-9._~-]+$", RegexOptions.CultureInvariant);

    private static string? VersionFromOfficial(Uri url)
    {
        if (!(IsPlainHttps(url) || IsSignedUpgrade(url)) || url.Host != Sources.UpgradeHost) return null;
        var match = Regex.Match(url.AbsolutePath, @"^/upgrade/client/win32-x86_64/spotify_installer-(1\.\d+\.\d+\.\d+\.g[0-9a-fA-F]+)-[0-9]+\.exe$", RegexOptions.CultureInvariant);
        return match.Success ? match.Groups[1].Value : null;
    }

    /// <summary>
    /// Turns typed input into an installable choice: a full version such as 1.2.80.699.gd5f6ebe3
    /// (served by the mirror's stable filename) or a link on Spotify's or the mirror's host.
    /// The version and signature are still verified after download, so a wrong guess fails safely.
    /// </summary>
    public static SpotifyChoice? TryCustom(string? text)
    {
        var input = text?.Trim();
        if (string.IsNullOrEmpty(input)) return null;
        if (FullVersionPattern().IsMatch(input))
            return new(input, MirrorFor(input), Custom: true);
        if (!Uri.TryCreate(input, UriKind.Absolute, out var url) || !(IsPlainHttps(url) || IsSignedUpgrade(url))) return null;
        if (url == Sources.LatestSpotify) return SpotifyChoice.Latest;
        if (VersionFromOfficial(url) is { } official) return new(official, url, Mirror: MirrorFor(official), Custom: true);
        if (VersionFromArchive(url) is { } archived) return new(archived, url, Mirror: MirrorFor(archived), Custom: true);
        var mirror = Regex.Match(url.AbsolutePath, @"^/download/spotify_installer-(1\.\d+\.\d+\.\d+\.g[0-9a-fA-F]+)-x64\.exe$", RegexOptions.CultureInvariant);
        return url.Host == Sources.MirrorHost && mirror.Success ? new(mirror.Groups[1].Value, url, Custom: true) : null;
    }

    /// <summary>
    /// Reads every Windows x64 release build. Accepts the LoadSpot table (one url per build), the legacy
    /// links schema, and the Pages feed, which may add "official", "archive" and "mirror" links, the
    /// "sha256" CI took from Spotify's file, and the "etag" that ties the permanent URL to this build.
    /// </summary>
    public static CatalogResult Read(string json, string tested)
    {
        var testedVersion = Parse(tested);
        using var document = JsonDocument.Parse(json);
        if (document.RootElement.ValueKind != JsonValueKind.Object)
            throw new InvalidDataException("LoadSpot returned an invalid version catalog.");
        var releases = new List<SpotifyChoice>();
        foreach (var property in document.RootElement.EnumerateObject())
        {
            var entry = property.Value;
            var full = Text(entry, "fullversion");
            if (full is null || !VersionPattern().IsMatch(full)) continue;
            if (!Version.TryParse(string.Join('.', full.Split('.').Take(4)), out var number)) continue;
            if (number.ToString() != property.Name) continue;
            var build = Text(entry, "buildType");
            if (build is not null && !build.Equals("Release", StringComparison.OrdinalIgnoreCase)) continue;
            var x64 = Child(Child(entry, "win"), "x64");
            var etag = Text(x64, "etag");
            var sha256 = Text(x64, "sha256") is { Length: 64 } hash && hash.All(Uri.IsHexDigit) ? hash.ToLowerInvariant() : null;
            var links = new[] { Text(x64, "official"), x64.ValueKind == JsonValueKind.String ? x64.GetString() : Text(x64, "url"),
                    Text(x64, "archive"), Text(x64, "mirror"), Text(Child(Child(entry, "links"), "win"), "x64") }
                .Select(link => Uri.TryCreate(link, UriKind.Absolute, out var url) ? url : null)
                .Where(url => url is not null).Distinct().ToList();
            // The permanent URL serves whatever build is current; without the ETag the feed saw, it cannot be pinned to this one.
            var permanent = etag is null ? null : links.FirstOrDefault(url => url == Sources.LatestSpotify);
            var official = permanent ?? links.FirstOrDefault(url => string.Equals(VersionFromOfficial(url!), full, StringComparison.OrdinalIgnoreCase));
            var archive = links.FirstOrDefault(url => IsArchive(url!, full));
            var mirror = links.FirstOrDefault(url => IsMirror(url!, full));
            if (official is null && archive is null && mirror is null) continue;
            // Spotify expires its versioned links, so a build known only by such a link still gets
            // the mirror's stable filename as a fallback; a missing mirror file fails like any 404.
            mirror ??= SpotifyVersions.MirrorFor(full);
            var sizeElement = Child(x64, "size");
            long size = sizeElement.ValueKind == JsonValueKind.Number && sizeElement.TryGetInt64(out var bytes) ? Math.Max(0, bytes) : 0;
            var date = Text(x64, "date");
            if (date is not null && !DateTime.TryParseExact(date, "dd.MM.yyyy", CultureInfo.InvariantCulture, DateTimeStyles.None, out _)) date = null;
            var primary = official ?? archive ?? mirror;
            releases.Add(new(full, primary, date, size, number == testedVersion, primary == mirror ? null : mirror, Archive: primary == archive ? null : archive,
                Sha256: sha256, ETag: primary == permanent ? etag : null));
        }
        releases.Sort((a, b) => Parse(b.FullVersion!).CompareTo(Parse(a.FullVersion!)));
        SpotifyChoice[] choices = [SpotifyChoice.Latest, .. releases];
        var selected = releases.FirstOrDefault(r => r.Recommended);
        return new(tested, choices, selected ?? SpotifyChoice.Latest,
            releases.Count == 0 ? "The catalog has no Windows x64 versions. The latest official installer is still available." :
            selected is null ? $"The catalog does not list the tested version {tested}. Its saved link is still available." : null);
    }

    private static JsonElement Child(JsonElement element, string property) =>
        element.ValueKind == JsonValueKind.Object && element.TryGetProperty(property, out var result) ? result : default;
    private static string? Text(JsonElement element, string property) =>
        Child(element, property) is { ValueKind: JsonValueKind.String } child ? child.GetString() : null;
}

public sealed class CatalogService(Downloads downloads)
{
    public async Task<CatalogResult> LoadAsync(CancellationToken token)
    {
        // Browsing does not depend on the patch server. Installation validates its current config separately.
        string? failure = null;
        foreach (var (source, seconds) in new[] { (Sources.CatalogApi, 4), (Sources.Catalog, 8) })
        {
            using var deadline = CancellationTokenSource.CreateLinkedTokenSource(token);
            deadline.CancelAfter(TimeSpan.FromSeconds(seconds));
            try
            {
                var result = SpotifyVersions.Read(await downloads.TextAsync(source, deadline.Token), Compatibility.TestedVersion);
                if (result.Choices.Count < 2) throw new InvalidDataException("No usable versions in the catalog.");
                return result;
            }
            catch (Exception error) when (error is HttpRequestException or IOException or InvalidDataException or JsonException or OperationCanceledException && !token.IsCancellationRequested)
            { failure = error.Message; }
        }
        token.ThrowIfCancellationRequested();
        return new(Compatibility.TestedVersion, [Compatibility.TestedChoice, SpotifyChoice.Latest], Compatibility.TestedChoice,
            $"Version list unavailable. The saved link for the tested version is still available. {failure}");
    }
}
